openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+            self._paths = paths
+
+            # Memoize the clone
+            memo[self] = clone
+
+        return memo[self]
+
+    def _get_nuclide_xml(self, nuclide, distrib=False):
+        xml_element = ET.Element("nuclide")
+        xml_element.set("name", nuclide[0])
+
+        if not distrib:
+            if nuclide[2] == 'ao':
+                xml_element.set("ao", str(nuclide[1]))
+            else:
+                xml_element.set("wo", str(nuclide[1]))
+
+        return xml_element
+
+    def _get_macroscopic_xml(self, macroscopic):
+        xml_element = ET.Element("macroscopic")
+        xml_element.set("name", macroscopic)
+
+        return xml_element
+
+    def _get_nuclides_xml(self, nuclides, distrib=False):
+        xml_elements = []
+        for nuclide in nuclides:
+            xml_elements.append(self._get_nuclide_xml(nuclide, distrib))
+        return xml_elements
+
+    def to_xml_element(self, cross_sections=None):
+        """Return XML representation of the material
+
+        Parameters
+        ----------
+        cross_sections : str
+            Path to an XML cross sections listing file
+
+        Returns
+        -------
+        element : xml.etree.ElementTree.Element
+            XML element containing material data
+
+        """
+
+        # Create Material XML element
+        element = ET.Element("material")
+        element.set("id", str(self._id))
+
+        if len(self._name) > 0:
+            element.set("name", str(self._name))
+
+        if self._depletable:
+            element.set("depletable", "true")
+
+        if self._volume:
+            element.set("volume", str(self._volume))
+
+        # Create temperature XML subelement
+        if self.temperature is not None:
+            element.set("temperature", str(self.temperature))
+
+        # Create density XML subelement
+        if self._density is not None or self._density_units == 'sum':
+            subelement = ET.SubElement(element, "density")
+            if self._density_units != 'sum':
+                subelement.set("value", str(self._density))
+            subelement.set("units", self._density_units)
+        else:
+            raise ValueError('Density has not been set for material {}!'
+                             .format(self.id))
+
+        if not self._convert_to_distrib_comps:
+            if self._macroscopic is None:
+                # Create nuclide XML subelements
+                subelements = self._get_nuclides_xml(self._nuclides)
+                for subelement in subelements:
+                    element.append(subelement)
+            else:
+                # Create macroscopic XML subelements
+                subelement = self._get_macroscopic_xml(self._macroscopic)
+                element.append(subelement)
+
+        else:
+            subelement = ET.SubElement(element, "compositions")
+
+            comps = []
+            allnucs = self._nuclides
+            dist_per_type = allnucs[0][2]
```

Comment:
the `[0][2]` seems magical - partly will go away if you introduce namedtuple for nuclides, but a comment about this (I think just say you compare allnucs to the first one to check for equality)